bpf/bpf_syscall.go

```diff
@@ -390,32 +379,76 @@ func GetMapNextKey(mapFD MapFD, k []byte, keySize int) ([]byte, error) {
 		return nil, err
 	}
 
-	bpfAttr := C.bpf_attr_alloc()
-	defer C.free(unsafe.Pointer(bpfAttr))
+	x := allocBpfSyscallMapMemCtx(keySize, 0)
+	defer x.Free()
 
+	return x.GetMapNextKey(mapFD, k, keySize)
+}
+
+type bpfSyscallMapMemCtx struct {
+	cK   unsafe.Pointer
+	cV   unsafe.Pointer
+	attr *[72]byte
+}
+
+func allocBpfSyscallMapMemCtx(kSize, vSize int) bpfSyscallMapMemCtx {
+	x := bpfSyscallMapMemCtx{
+		cK:   C.malloc(C.size_t(kSize)),
+		attr: C.bpf_attr_alloc(),
+	}
+
+	if vSize > 0 {
+		x.cV = C.malloc(C.size_t(vSize))
+	}
+
+	return x
+}
+
+func (x bpfSyscallMapMemCtx) Free() {
+	C.free(x.cK)
+	C.free(x.cV)
+	C.free(unsafe.Pointer(x.attr))
+}
+
+func (x bpfSyscallMapMemCtx) GetMapNextKey(mapFD MapFD, k []byte, keySize int) ([]byte, error) {
 	var cK unsafe.Pointer
 
 	// Have to make C-heap copies here because passing these to the syscalls is done via pointers in an
 	// intermediate struct.
 	//
 	// it is valid to pass a nil key to start the iteration
-	if k != nil {
-		cK = C.CBytes(k)
-		defer C.free(cK)
+	if len(k) >= 1 {
+		cK = unsafe.Pointer(&k[0])
 	}
-	cV := C.malloc(C.size_t(keySize)) // value has the size of a key - it is the key!
-	defer C.free(cV)
 
-	C.bpf_attr_setup_map_get_next_key(bpfAttr, C.uint(mapFD), cK, cV, 0)
+	C.bpf_attr_clear(x.attr)
+	C.bpf_attr_setup_map_get_next_key(x.attr, C.uint(mapFD), cK, x.cK, 0)
```

Comment:
Assuming k is allocated as a Go slice, I don't think it's valid to embed a pointer to k[0] in the attr union.  It breaks one of the pointer rules in the [cgo docs](https://golang.org/cmd/cgo/#hdr-Passing_pointers):

> C code may not keep a copy of a Go pointer after the call returns.

and

> Go code may not store a Go pointer in C memory. C code may store Go pointers in C memory, subject to the rule above: it must stop storing the Go pointer when the C function returns.

Reply:
Hmmm my understanding was that this does not fall in that cathegory as C technically does not keep that pointer and go still has a live pointer, but you are likely right that letting one function to fill it in and another function to make the syscall may violate it. I will unify it so that it does not split in 2 calls.